[SPARK-10180] [SQL] JDBC datasource are not processing EqualNullSafe filter

HyukjinKwon opened the pull request:
This PR is followed by https://github.com/apache/spark/pull/8391.
Previous PR fixes JDBCRDD to support null-safe equality comparison for JDBC datasource. This PR fixes the problem that it can actually return null as a result of the comparison resulting error as using the value of that comparison.

HyukjinKwon committed:
[SPARK-10180] [SQL] Update query to prevent to return null.

AmplabJenkins commented:
Can one of the admins verify this patch?

JoshRosen commented:
Please update the pull request description to describe the changes in this patch.

JoshRosen commented:
Will this case be covered by the end-to-end docker tests?

HyukjinKwon commented:
Should I better write a test code for this?

HyukjinKwon commented:
Should I better write some test codes for this?

AmplabJenkins commented:
Can one of the admins verify this patch?

yhuai commented:
Can we add a test?

HyukjinKwon committed:
[SPARK-10180][SQL] Add a test for the null-safety comparison operator.

HyukjinKwon commented:
I added a simple test for this. I wanted to add a test including `null` but when the given value is `null`, Spark converts it to `IsNull`. To make this worse, more complicated logics cannot be applied here to test null-safety properly since now filters are pushed down only for basic operations (not even cast or arithmetic calculations). For now, I think this is the best to test this.

HyukjinKwon commented:
test this please

HyukjinKwon commented:
It looks Jenkins does not run the test for the past commits that I made as a user not added to whitelist. Would anybody please run the test for this please if it looks good?

yhuai commented:
test this please

SparkQA commented:
**[Test build #46425 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46425/consoleFull)** for PR 8743 at commit [`bf0b22c`](https://github.com/apache/spark/commit/bf0b22c75b1f5084a07daa903d554dafc20d9a6a).

SparkQA commented:
**[Test build #46425 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46425/consoleFull)** for PR 8743 at commit [`bf0b22c`](https://github.com/apache/spark/commit/bf0b22c75b1f5084a07daa903d554dafc20d9a6a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46425/
Test PASSed.

HyukjinKwon commented:
Just a question. Now it looks the PR for the end-to-end docker tests is merged. Do you think it needs all the tests for all the databases (namely  `MySQLIntegrationSuite` and `PostgresIntegrationSuite` for now)?

HyukjinKwon committed:
Resolve conflicts

HyukjinKwon committed:
Resolve conflicts

HyukjinKwon committed:
Use != instead of <>.

HyukjinKwon committed:
Remove unintentionally added functions of old verions

rxin commented:
@zsxwing can you review this one?

maropu commented:
@HyukjinKwon +1, I think it'd better to add tests in `MySQLIntegrationSuite` and `PostgresIntegrationSuite`.

HyukjinKwon commented:
Hm.. actually I think adding tests more in docker is a bit over-tested. It looks the comparison operators I used are all already being used in `compileFilter` and i think the logical operators can be covered in the existing test.

maropu commented:
Yes and you're right. We need to clearly define the rule that decides which tests should be added in `docker-integration-tests`.

zsxwing commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala:
```diff
@@ -191,6 +191,9 @@ private[sql] object JDBCRDD extends Logging {
   private def compileFilter(f: Filter): String = f match {
     case EqualTo(attr, value) => s"$attr = ${compileValue(value)}"
+    case EqualNullSafe(attr, value) =>
+      s"(NOT ($attr != ${compileValue(value)} OR $attr IS NULL OR " +
```
It took me several minutes to understand it. Maybe `(attr IS NULL AND value IS NULL) OR (attr <> null AND value <> null AND attr = value)` is easier.

zsxwing replied:
Ah, sorry, my previous expr doesn't work.

zsxwing commented:
I'm wondering if it's okey to add `sealed` to `Filter` and remove `case _`. If so, the compiler can help us find such issue.

HyukjinKwon commented:
@zsxwing I will anyway resolve the conflicts first

HyukjinKwon committed:
Resolve conflicts

rxin commented:
Sorry I kept merging pull requests that made your life harder. Can you bring it up to date again? I promise this is the last one.

HyukjinKwon committed:
Resolve conflicts

Conflicts:
	sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala

HyukjinKwon commented:
@rxin Nothing is eazy and happy new year. I just resolved conflicts.

HyukjinKwon commented:
test this please

HyukjinKwon commented:
Could anybody please type "test this please" for this PR? I can't trigger a test for this PR.

SparkQA commented:
**[Test build #2287 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2287/consoleFull)** for PR 8743 at commit [`17b6953`](https://github.com/apache/spark/commit/17b695334e07ec9a58e766bd737d1c706f714109).

SparkQA commented:
**[Test build #2287 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2287/consoleFull)** for PR 8743 at commit [`17b6953`](https://github.com/apache/spark/commit/17b695334e07ec9a58e766bd737d1c706f714109).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

HyukjinKwon committed:
Change the compiled string test.

HyukjinKwon commented:
test this please

SparkQA commented:
**[Test build #2291 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2291/consoleFull)** for PR 8743 at commit [`a489a33`](https://github.com/apache/spark/commit/a489a336b6ef4364f834262738f877596b67297c).

SparkQA commented:
**[Test build #2291 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2291/consoleFull)** for PR 8743 at commit [`a489a33`](https://github.com/apache/spark/commit/a489a336b6ef4364f834262738f877596b67297c).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #2294 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2294/consoleFull)** for PR 8743 at commit [`a489a33`](https://github.com/apache/spark/commit/a489a336b6ef4364f834262738f877596b67297c).

SparkQA commented:
**[Test build #2294 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2294/consoleFull)** for PR 8743 at commit [`a489a33`](https://github.com/apache/spark/commit/a489a336b6ef4364f834262738f877596b67297c).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

HyukjinKwon commented:
test this please

SparkQA commented:
**[Test build #2295 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2295/consoleFull)** for PR 8743 at commit [`a489a33`](https://github.com/apache/spark/commit/a489a336b6ef4364f834262738f877596b67297c).

SparkQA commented:
**[Test build #2295 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2295/consoleFull)** for PR 8743 at commit [`a489a33`](https://github.com/apache/spark/commit/a489a336b6ef4364f834262738f877596b67297c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

rxin commented:
Thanks - I've merged this.
